cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,352 @@
+        return self._description
+
+    def predicate(self, gate: raw_types.Gate) -> bool:
+        """Checks whether `cirq.Gate` instance `g` belongs to this GateFamily.
+
+        The default predicate depends on the gate family initialization type:
+            a) Instance Family: `gate == self.gate`.
+            b) Type Family: `isinstance(gate, self.gate)`.
+
+        Args:
+            gate: `cirq.Gate` instance which should be checked for containment.
+        """
+        return (
+            gate == self.gate
+            if isinstance(self.gate, raw_types.Gate)
+            else isinstance(gate, self.gate)
+        )
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        if isinstance(item, raw_types.Operation):
+            if item.gate is None:
+                return False
+            item = item.gate
+        return self.predicate(item)
+
+    def __str__(self) -> str:
+        return f'{self.name}\n{self.description}'
+
+    def __repr__(self) -> str:
+        return (
+            f'cirq.GateFamily(gate={self._gate_str(repr)},'
+            f'name="{self.name}", '
+            f'description="{self.description}")'
+        )
+
+    def _value_equality_values_(self) -> Any:
+        # `isinstance` is used to ensure the a gate type and gate instance is not compared.
+        return isinstance(self.gate, raw_types.Gate), self.gate, self.name, self.description
+
+
+@value.value_equality()
+class Gateset:
+    """Gatesets represent a collection of `cirq.GateFamily` objects.
+
+    Gatesets are useful for
+        a) Describing the set of allowed gates in a human readable format
+        b) Validating a given gate / optree against the set of allowed gates
+
+    Gatesets rely on the underlying `cirq.GateFamily` for both description and
+    validation purposes.
+    """
+
+    def __init__(
+        self,
+        *gates: Union[Type[raw_types.Gate], raw_types.Gate, GateFamily],
+        name: Optional[str] = None,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
+    ) -> None:
+        """Init Gateset.
+
+        Accepts a list of gates, each of which should be either
+            a) `cirq.Gate` subclass
+            b) `cirq.Gate` instance
+            c) `cirq.GateFamily` instance
+
+        `cirq.Gate` subclasses and instances are converted to the default
+        `cirq.GateFamily(gate=g)` instance and thus a default name and
+        description is populated.
+
+        Args:
+            *gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
+                validated by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, `cirq.GlobalPhaseOperation` is accepted.
```

Comment:
Looking closer at this boolean, do we need to special case this here. Wouldn't it make for a simpler API if the user just included a `cirq.GateFamily(cirq.GlobalPhaseOperation)` (type based) when constructing the `GateSet` they want to use instead of having this boolean flag for things ?

Reply:
Well, constructing a `GateFamily` with an operation that doesn't have any underlying "gate" can also be pretty confusing? 

I would prefer not to construct `GateFamilies` that work on any operation type. It opens up a new can of worms and would bring the all the complexities and confusion of Gates vs Operations relationships to the Gateset world as well. 

This is also the reason why `__contains__` in a gateset should be simple. We can fix the the above mentioned confusion using documentation -- use `gateset.validate()` for validating objects. Use containment for checking whether a "gate" (or a "gate" applied to qubits) is contained in the "gateset" or not.

Reply:
Ahhh god point. Makes a lot of sense to keep here then.